gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAChallengeEndpoint.java

```diff
@@ -190,17 +206,56 @@ private void verifyCode(RoutingContext routingContext) {
         factorData.put(FactorContext.KEY_ENROLLED_FACTOR, enrolledFactor);
         factorData.put(FactorContext.KEY_CODE, code);
         factorData.put(FactorContext.KEY_REQUEST, new EvaluableRequest(new VertxHttpServerRequest(routingContext.request().getDelegate())));
+
+        if(factor.is(FIDO2)){
+            factorData.put(ConstantKeys.PASSWORDLESS_CHALLENGE_KEY, routingContext.session().get(PASSWORDLESS_CHALLENGE_KEY));
+            factorData.put(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY, routingContext.session().get(PASSWORDLESS_CHALLENGE_USERNAME_KEY));
+            factorData.put(ConstantKeys.PASSWORDLESS_ORIGIN, domain.getWebAuthnSettings().getOrigin());
+        }
+
         final FactorContext factorCtx = new FactorContext(applicationContext, factorData);
 
         verify(factorProvider, factorCtx, h -> {
             if (h.failed()) {
                 handleException(routingContext);
                 return;
             }
-            // save enrolled factor if needed and redirect to the original url
-            final MultiMap queryParams = RequestUtils.getCleanedQueryParams(routingContext.request());
-            final String returnURL = getReturnUrl(routingContext, queryParams);
 
+            if (factor.is(FIDO2)) {
+                final String userId = routingContext.session().get(ConstantKeys.PASSWORDLESS_CHALLENGE_USER_ID);
+                final JsonObject webauthnResp = new JsonObject(code);
+                final String credentialId = webauthnResp.getString("id");
+                updateCredential(routingContext.request(), credentialId, userId, ch -> {
+
+                    if (ch.failed()) {
+                        final String username = routingContext.session().get(PASSWORDLESS_CHALLENGE_USERNAME_KEY);
+                        logger.error("An error has occurred while authenticating user {}", username, h.cause());
```

Comment:
An error has occurred while updating a credential for user {}

Reply:
done